### PR TITLE
[IMP][9.0] base_external_dbsource: Refactor & Split by source

### PR DESCRIPTION
**Pull request by lasley**
This is a refactoring of the `base_external_dbsource` module, hopefully allowing for easier addition and more isolation of the independent data sources.

Note that all adapter modules are `auto_install` to `True` in order to not break existing installations while still allowing for the modules to be split logically. In v11 we can then add the proper Python depends, which are already in the manifests but commented out.

Due to the above note, I also had to submit all of the modules in one PR. Separating them will break things horribly. Sorry for the gigantic PR

Depends:
- [x] #637

### CODE REVIEW

**Comment by lasley** [comment]
Alright I'm done here. Not sure what's up with Runbot - it's acting like it's not installing the FDB from requirements.txt, but the error isn't replicating on Travis. I'll probably have to look more into that, but this can be reviewed with that & the license change #637 in mind.

File: base_external_dbsource/README.rst [hunk 1]
```diff
@@ -46,15 +45,19 @@ To use this module:
 Known issues / Roadmap
 ======================
 
+* Find a way to remove or default the CA certs dir
+* Add concept of multiple connection strings for one source (multiple nodes)
+* Add a ConnectionEnvironment that allows for the reuse of connections
+* Message box should be displayed instead of error in ``connection_test``
+* Remove old api compatibility layers (v11)
```
Comment (lasley) [inline base_external_dbsource/README.rst:52]: `* Add CRUD interface (use Elasticsearch module additions as template)`

Comment (lasley) [reply base_external_dbsource/README.rst:52]: Implemented - still needs tests

**Comment by lasley** [comment]
Alright this is good to review; now including a fancy new CRUD interface if adapters should choose to implement them & rebased on 9.0 to fix requirements conflict.

**Comment by lasley** [comment]
Rebased now that dependency is merged

**Comment by lasley** [comment]
@t3ddftw please review

File: base_external_dbsource/README.rst [hunk 2]
```diff
@@ -46,15 +45,20 @@ To use this module:
...
 Bug Tracker
 ===========
 
 Bugs are tracked on `GitHub Issues <https://github.com/OCA/server-tools/issues>`_.
 In case of trouble, please check there if your issue has already been reported.
-If you spotted it first, help us smashing it by providing a detailed and welcomed feedback `here <https://github.com/OCA/
-server-tools/issues/new?body=module:%20
-base_external_dbsource%0Aversion:%20
-9.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback.
```
Comment (tedsalmon) [inline base_external_dbsource/README.rst:61]: s/smashing/smash

File: base_external_dbsource/tests/test_base_external_dbsource.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        kwargs = {'kwargs': True}
+        self.dbsource.current_table = 'table'
+        res, adapter = self._test_adapter_method(
+            'remote_browse', create=True, args=args, kwargs=kwargs,
+        )
+        adapter.assert_called_once_with(*args, **kwargs)
+        self.assertEqual(res, adapter())
+
+    def test_remote_browse_asserts_current_table(self):
+        """ It should raise AssertionError if a table not selected """
```
Comment (tedsalmon) [inline base_external_dbsource/tests/test_base_external_dbsource.py]: s/if a table not/if a table is not

File: base_external_dbsource/tests/test_base_external_dbsource.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        kwargs = {'kwargs': True}
+        self.dbsource.current_table = 'table'
+        res, adapter = self._test_adapter_method(
+            'remote_create', create=True, args=args, kwargs=kwargs,
+        )
+        adapter.assert_called_once_with(*args, **kwargs)
+        self.assertEqual(res, adapter())
+
+    def test_remote_create_asserts_current_table(self):
+        """ It should raise AssertionError if a table not selected """
```
Comment (tedsalmon) [inline base_external_dbsource/tests/test_base_external_dbsource.py]: s/table not/table is not

File: base_external_dbsource/tests/test_base_external_dbsource.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        kwargs = {'kwargs': True}
+        self.dbsource.current_table = 'table'
+        res, adapter = self._test_adapter_method(
+            'remote_delete', create=True, args=args, kwargs=kwargs,
+        )
+        adapter.assert_called_once_with(*args, **kwargs)
+        self.assertEqual(res, adapter())
+
+    def test_remote_delete_asserts_current_table(self):
+        """ It should raise AssertionError if a table not selected """
```
Comment (tedsalmon) [inline base_external_dbsource/tests/test_base_external_dbsource.py]: s/table not/table is not

File: base_external_dbsource/tests/test_base_external_dbsource.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+        kwargs = {'kwargs': True}
+        self.dbsource.current_table = 'table'
+        res, adapter = self._test_adapter_method(
+            'remote_search', create=True, args=args, kwargs=kwargs,
+        )
+        adapter.assert_called_once_with(*args, **kwargs)
+        self.assertEqual(res, adapter())
+
+    def test_remote_search_asserts_current_table(self):
+        """ It should raise AssertionError if a table not selected """
```
Comment (tedsalmon) [inline base_external_dbsource/tests/test_base_external_dbsource.py]: s/table not/table is not

File: base_external_dbsource/tests/test_base_external_dbsource.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+        kwargs = {'kwargs': True}
+        self.dbsource.current_table = 'table'
+        res, adapter = self._test_adapter_method(
+            'remote_update', create=True, args=args, kwargs=kwargs,
+        )
+        adapter.assert_called_once_with(*args, **kwargs)
+        self.assertEqual(res, adapter())
+
+    def test_remote_update_asserts_current_table(self):
+        """ It should raise AssertionError if a table not selected """
```
Comment (tedsalmon) [inline base_external_dbsource/tests/test_base_external_dbsource.py]: s/table not/table is not

File: base_external_dbsource_firebird/models/base_external_dbsource.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+# Copyright 2011 Daniel Reis
+# Copyright 2016 LasLabs Inc.
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+
+import logging
+
+from openerp import api
+from openerp import models
```
Comment (tedsalmon) [inline base_external_dbsource_firebird/models/base_external_dbsource.py]: Condense this to a single line, IMO.

File: base_external_dbsource_firebird/models/base_external_dbsource.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        CONNECTORS.append(('fdb', 'Firebird'))
+    except:
+        _logger.info('Firebird library not available. Please install "fdb" '
+                     'python package.')
+except ImportError:
+    _logger.info('base_external_dbsource Odoo module not found.')
+
+
+class BaseExternalDbsource(models.Model):
+    """ It provides logic for connection to an Firebird data source. """
```
Comment (tedsalmon) [inline base_external_dbsource_firebird/models/base_external_dbsource.py]: s/an/a

File: base_external_dbsource_mssql/README.rst [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+======================
+
+* Add X.509 authentication
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/server-tools/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback.
```
Comment (tedsalmon) [inline base_external_dbsource_mssql/README.rst:54]: s/smashing/smash

**Review by tedsalmon — APPROVED** [review]
LGTM! 👍 

My review items are all just spelling :p

File: base_external_dbsource/README.rst (hunk 2 above)
Comment (lasley) [reply base_external_dbsource/README.rst:61]: This is part of the template, and while I do agree the wording needs to be changed, it is out of scope here. Proposing a change to the template for this is something I've been meaning to do, you're welcome to as well - https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

File: base_external_dbsource/tests/test_base_external_dbsource.py (hunk 3 above)
Comment (lasley) [reply base_external_dbsource/tests/test_base_external_dbsource.py]: (you can save yourself some effort and just tell me to apply X change to all instances)

**Comment by lasley** [comment]
Thanks for the review @t3ddftw - most points fixed, some responded to.

File: base_external_dbsource_mssql/README.rst (hunk 10 above)
Comment (lasley) [inline base_external_dbsource_mssql/README.rst]: Really it should be `to smash`

File: base_external_dbsource/README.rst (hunk 2 above)
Comment (lasley) [reply base_external_dbsource/README.rst:61]: OCA/maintainer-tools#251

**Comment by lasley** [comment]
Added build fix for OCA/maintainer-quality-tools#427

**Pull request closed without merging**